src/Avalonia.Controls/Control.cs

```diff
@@ -215,18 +279,89 @@ void IVisualBrushInitialize.EnsureInitialized()
         /// <returns>The control that receives the focus adorner.</returns>
         protected virtual IControl? GetTemplateFocusTarget() => this;
 
+        /// <summary>
+        /// Invoked as the first step of marking the control as loaded and raising the
+        /// <see cref="Loaded"/> event.
+        /// </summary>
+        internal void OnLoadedCore()
+        {
+            if (_isLoaded == false)
+            {
+                _isLoaded = true;
+                OnLoaded();
+            }
+        }
+
+        /// <summary>
+        /// Invoked as the first step of marking the control as unloaded and raising the
+        /// <see cref="Unloaded"/> event.
+        /// </summary>
+        internal void OnUnloadedCore()
+        {
+            if (_isLoaded)
+            {
+                _isLoaded = false;
+                OnUnloaded();
+            }
+        }
+
+        /// <summary>
+        /// Invoked just before the <see cref="Loaded"/> event.
+        /// </summary>
+        protected virtual void OnLoaded()
+        {
+            var eventArgs = new RoutedEventArgs(LoadedEvent);
+            eventArgs.Source = null;
+            RaiseEvent(eventArgs);
+        }
+
+        /// <summary>
+        /// Invoked just before the <see cref="Unloaded"/> event.
+        /// </summary>
+        protected virtual void OnUnloaded()
+        {
+            var eventArgs = new RoutedEventArgs(UnloadedEvent);
+            eventArgs.Source = null;
+            RaiseEvent(eventArgs);
+        }
+
         /// <inheritdoc/>
         protected sealed override void OnAttachedToVisualTreeCore(VisualTreeAttachmentEventArgs e)
         {
             base.OnAttachedToVisualTreeCore(e);
 
             InitializeIfNeeded();
+
+            // Note: This code is not multi-threaded ready and assumes to run only on the UIThread
+            if (_isLoaded == false)
+            {
+                _loadedQueue.Add(this);
+
+                if (_isLoadedProcessingRequested == false)
```

Comment:
We can check the return value from `_loadedQueue.Add(this);` to know if control `Loaded` event is already scheduled. We also should not call `Dispatcher.UIThread.Post` for every control. The handler and flag should be both static, so we would schedule callback only once.

Reply:
The Flag should be static, yes. I did this too late at night I guess. I'm not sure the handler needs to be as the flag prevents it from being built unless needed. However, I suppose if static it would be more efficient as it isn't rebuilding the code all the time.

And yes, we don't need to call the dispatcher if control was already added. Good optimization.

Reply:
Updated